Introduce debug_increaseTime RPC method

albrow opened the pull request:
We are using Geth to tests our smart contracts in a single-node private testnet. Some of those contracts have time dependencies (e.g., a funds can be released only after a certain amount of time). The ability to increase time artificially greatly improves the performance and reliability of our tests.

This PR introduces a new RPC method `debug_increaseTime` which has the effect of increasing the timestamp of the next mined block. It's modeled after the [`evm_increaseTime` endpoint in Ganache](https://github.com/trufflesuite/ganache-cli#implemented-methods) and behaves similarly. The endpoint takes a single parameter, an offset to apply to an adjustable clock in seconds. Internally, this is implemented via a new "aclock" package.

Time can only be increased, and can never be decreased or the offset restored to its previous value. The reason is that I expect a lot code would break (or goroutines would stall) if we went backwards in time.

Our smart contract tests utilize this new RPC method right now (running on our forked version of Geth) and it seems to work. I would also be willing to write some tests for the new method in go-ethereum itself, but would need some guidance on the best way to do that.

albrow commented:
@karalabe or @holiman this PR has been open for a long time and I haven't received any feedback. We are making extensive use of the `debug_increaseTime` feature at [0x](https://github.com/0xProject/0x-monorepo) and we would love to get this merged upstream if possible.

Is this a feature you would consider merging? Any thoughts/feedback/questions?

albrow commented:
Thanks @holiman! I'll fix these merge conflicts as soon as I get a chance.

fjl commented:
The functionality is nice. The implementation in this PR is essentially a global variable with a fancy API on top, used in a couple of places but not others. IMHO it reaches too far into internals. It would be better to add the time offset functionality to the clique consensus engine only.

fjl commented:
@albrow Are you using clique for testing?

albrow commented:
@fjl Yes, we are using Clique for testing at 0x. I could probably remove `aclock` from the ethhash package since we don't use that (I imagine most people using Geth for testing similar to how we are are using Clique). However, I don't think we can keep the changes isolated to just the clique package.

Take for example, this excerpt from [ __core/blockchain.go__](https://github.com/ethereum/go-ethereum/blob/f278e20b4ae4d7f173153d0bb2ca5b81b657cc20/core/blockchain.go#L1080):

```
// Allow up to MaxFuture second in the future blocks. If this limit is exceeded
// the chain is discarded and processed at a later time if given.
max := big.NewInt(time.Now().Unix() + maxTimeFutureBlocks)
```

IIRC I needed to change `time.Now()` to `aclock.Now()` here in order to prevent an issue where new blocks won't be mined due to being perceived as too far into the future. It has been a while though, so I can reconfirm this.

albrow commented:
@fjl I rebased and updated this PR and reduced the number of places where `aclock` is used. 

I'm assuming based on your feedback that we only want to support increasing the time for the Clique consensus engine. In order to do this, in addition to making changes to the clique package itself, I needed to make some changes to __miner/worker.go__. Without these changes, the miner will get stuck with a log message that looks like this:

```
INFO [01-29|14:50:06.886] Mining too far in the future             wait=277h46m41s
```

I believe the current version of the PR has the minimum number of changes needed. Let me know what you think!

nventuro commented:
Hello! It was great to see this PR move forward, are there any plans to have it merged soon? Not only would having multiple testing platforms be extremely useful, but Ganache also doesn't [support certain debug-related features](https://github.com/trufflesuite/ganache-core/issues/47) that are included in Geth. 

Additionally, tools such as [0x's profiling and code-coverage tools](https://blog.0xproject.com/new-ethereum-dev-tools-from-0x-db80ee9e802) have also been developed using Geth, and present multiple issues when attempting to have them run in e.g. Ganache.

As of today, the only thing holding back OpenZeppelin migrating to Geth-based testing is lack of support for time manipulation utilities in debug mode.

albrow commented:
@fjl still waiting for feedback from you.

pinkiebell commented:
Gentle ping, I have interests in this PR too 🙂

fjl commented:
@albrow Sorry for the very slow conversation on this PR. I think we're moving in the right direction here, but still need to change some things to make this work right.

My original concern about the global variable is not addressed so far. The recent changes have reduced the usage of package aclock, but the package still works the same way. I think we don't need this package at all. Instead, we should define the current time as a field in the clique engine. This could be done using package common/mclock. Since the miner needs to be aware of the clock, we'd need to add the clock to it as well.

albrow commented:
@fjl this PR has been open for 15 months. Given the lack of response, we have shifted our testing strategy at my company and are no longer relying on Geth for testing. I don't have time to work on this anymore. If this is still a feature you're interested in someone else will need to take over.
